### ISSUE TITLE
Cancelling a stream via a cancellation token can lead to incorrect results

### ISSUE THREAD

**Issue by jskeet** [grpc/grpc#17761]
### What version of gRPC and what language are you using?

1.17.1, C# 
 
### What operating system (Linux, Windows, …) and version?

Windows 
 
### What runtime / compiler are you using (e.g. python version or version of gcc)

.NET Core SDK 2.1.402
 
### What did you do?

Create a small but non-empty table in a Spanner database, then run the code below, providing the project ID, Spanner instance ID, Spanner database ID and table name.

```csharp
using Google.Cloud.Spanner.Common.V1;
using Google.Cloud.Spanner.V1;
using Grpc.Core;
using System;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static async Task Main(string[] args)
    {
        if (args.Length != 4)
        {
            Console.WriteLine($"Required arguments: <project ID> <instance ID> <database ID> <table name>");
            return;
        }
        TaskScheduler.UnobservedTaskException += (sender, e) => Console.WriteLine(e.Exception);
        var dbName = new DatabaseName(args[0], args[1], args[2]);
        var client = SpannerClient.Create();
        var grpcClient = client.GrpcClient;

        var session = await grpcClient.CreateSessionAsync(new CreateSessionRequest { DatabaseAsDatabaseName = dbName });

        try
        {
            var request = new ExecuteSqlRequest
            {
                SessionAsSessionName = session.SessionName,
                Sql = $"SELECT * from {args[3]}",
            };

            using (var streamingCall = grpcClient.ExecuteStreamingSql(request))
            {
                var cts = new CancellationTokenSource();
                cts.Cancel();
                // Comment out the following line to see cancellation work properly
                await streamingCall.ResponseHeadersAsync;
                int count = 0;
                while (await streamingCall.ResponseStream.MoveNext(cts.Token))
                {
                    Console.WriteLine($"Read response {++count}");
                }
                Log("Done");
            }
        }
        catch (RpcException e)
        {
            Log($"Caught RPC exception of status {e.StatusCode}");
        }
    }

    static void Log(string message) =>
        Console.WriteLine($"{DateTime.UtcNow:HH:mm:ss.FFFFFF}: {message}");
}
```

### What did you expect to see?

Either a cancellation exception *or* successfully reading a single row, if the cancellation isn't observed in time. Either of those is fine. 
 
### What did you see instead?
 
Just "Done" indicating it had *successfully* reached the end of the stream, having read 0 responses.

- If you comment out the cts.Cancel() call, you should see one response ("Read response 1").
- If you comment out the `await streamingCall.ResponseHeadersAsync` statement, cancellation is observed properly

It's just possible that this is a Spanner bug, of course - although arguably *any* call to `streamingCall.ResponseStream.MoveNext(cts.Token)` should return a failed task when `cts.Token` is already cancelled before the call, even if we've already seen the end of the stream.

**Comment by jtattermusch** [comment]
First results:

I am able to reproduce a variant of this issue:
1. once the RPC start, server quickly sends all its responses and status to the client.
2. a while later (one the call has actually finished on the serverside), the client request cancellation of the call (regardless of via responseStream.MoveNext(cancellationToken) or via call.Cancel()). 

this seems to mess with the way C-core delivers the responses (they are lost), but C-core still propagates the status received from the server up to the C# layer (the call on server side ended with Status.OK, so this is what gets reported to C#.

**Comment by jtattermusch** [comment]
@jskeet  can you confirm that replacing `await streamingCall.ResponseHeadersAsync` by `await Task.Delay(3000)` also reproduces the same issue? (which would support the theory described above, as waiting for response headers probably waits long enough so that all the responses are delivered to the client before MoveNext() starts which then creates the problem)?

**Comment by jskeet** [comment]
Confirmed - yes, that shows the same result.

**Comment by jtattermusch** [comment]
After some investigation, I have a lead what could be the problem.
It looks like that under the circumstances described above, C-core reports a problem with the read via setting the arg `success` to false (which is an unusual scenario). In C# code, we happen to treat that as a null message (thus ending the stream and expecting the call statusCode will be non-ok), but in this call the status' status received from the server is actually OK, so the call is treated as successful:
https://github.com/grpc/grpc/blob/343c2286af2bd31efa8277d4ef50f79b27490677/src/csharp/Grpc.Core/Internal/AsyncCallBase.cs#L343

**Comment by jtattermusch** [comment]
After not being able to sufficiently fix in the C# layer, I filed #18234 to investigate how this should be fixed in C core.

**Comment by jtattermusch** [comment]
Still a problem.

**Comment by tonydnewell** [comment]
It appears that this has been fixed a while back.

A version of the test in PR https://github.com/grpc/grpc/pull/18124 now passes (I had to change the exception that was caught).
See also https://github.com/grpc/grpc/issues/18234#issuecomment-578801130

I will create a new PR to add a test.

**Comment by tonydnewell** [comment]
Verified and added test in PR https://github.com/grpc/grpc/pull/30523

**Comment by jtattermusch** [comment]
> It appears that this has been fixed a while back.

We need to know when exactly and by what it got fixed (we can figure that out by e.g. eyeballing it or by git bisect).

> 
> A version of the test in PR #18124 now passes (I had to change the exception that was caught). See also [#18234 (comment)](https://github.com/grpc/grpc/issues/18234#issuecomment-578801130)

Can you confirm that you can reproduce with the repro on older codebase? (before some fix went in)?

> 
> I will create a new PR to add a test.

**Comment by tonydnewell** [comment]
I can narrow down the fix to:
- Fixed: Grpc 2.31-0-pre1 (tag  v1.31.0-pre1) 22 July 2020
- Broken: Grpc 2.30.0 (tag v1.30.0) 22 June 2020

I did this by testing against the nuget packages.

However I've not narrowed down any further with `git bisect` since the build tools/environment are different enough to fail to build in my current dev environment.  There are 705 commits between the 2 labels.

**Comment by jtattermusch** [comment]
https://github.com/grpc/grpc/pull/30523 added a test that verifies that this issue has indeed been fixed (and it looks like it's been fixed starting from ~2.31.0 release)
